### DISCUSSION TITLE
Assert() and free() of memory

### DISCUSSION THREAD

**Discussion by Hussain1811** [google/googletest#4562] [Community Help]
Hello,
I wanted to explain my problem via the below code:

```
int func(void *ptr)
{
//function logic
}

TEST()
{
   int rc = 0;
   void *ptr = NULL;
   ptr = malloc(10);
   ASSERT(ptr != NULL);
   rc = func(ptr);
   ASSERT(rc == 0);
   free(ptr);
}
```

So if my assert on `rc` fails, then the memory won't be freed.

How do i handle this situation?

**Comment by Shriharsh-Deshmukh** [comment]
To ensure memory is freed even if an assertion fails, you can use `ASSERT_NO_THROW` or a `try-catch` block, but a cleaner way in GTest is using `SCOPED_TRACE` along with `EXPECT` statements, or better yet, using a `defer`-like mechanism in C++. Here’s how you can handle it:

1. **Use `EXPECT` instead of `ASSERT`**: `EXPECT` continues the test execution, allowing you to reach the `free(ptr)` line.

2. **Use RAII (Resource Acquisition Is Initialization)**: A C++ class that manages the resource and automatically frees the memory when it goes out of scope is ideal.

Here's how to implement this:

```cpp
class MemoryManager {
public:
    explicit MemoryManager(void* ptr) : ptr_(ptr) {}
    ~MemoryManager() { free(ptr_); }

private:
    void* ptr_;
};

TEST(TestSuite, TestName) {
    int rc = 0;
    void *ptr = malloc(10);
    ASSERT_NE(ptr, nullptr);  // Check allocation

    // Ensure memory is freed automatically when out of scope
    MemoryManager mem_manager(ptr);

    rc = func(ptr);
    EXPECT_EQ(rc, 0);  // Continue execution even if this fails
    // No need for explicit free(ptr) as it will be handled by MemoryManager
}
```

This ensures that memory is freed even if the test fails.

**Comment by Hussain1811** [comment]
Thank you @Shriharsh-Deshmukh 
This makes it clear
